### DISCUSSION TITLE
CI: Protecting Semantic Release before merging

### DISCUSSION THREAD

**Discussion by oscard0m** [octokit/octokit.js#2018] [Ideas] [answered]
Hi all!

Opening a discussion / brainstorming on how we could improve our Pull Requests Semantic Commits. This discussion is inspired by the thread in this [issue](https://github.com/octokit/webhooks/pull/377#issuecomment-782898105).

It would be nice to find a way to prevent merging a PR with a commit (if squashed) or set of commits (if not squashed) accomplishing Semantic Commit rules to preserve and assure all the flows around [Semantic Release](https://github.com/semantic-release/semantic-release) in Octokit Ecosystem are archived.

At the same time, it would be important to remove friction to contributors when working in a PR and not forcing them to make all commits semantic if they plan to squash.

---

@G-Rath [come up with a proposal](https://github.com/octokit/webhooks/pull/377#issuecomment-782952300) here: https://github.com/octokit/webhooks/commit/8506555c633acfbea301af8fae57e9ce214d0e53

**Comment by oscard0m** [comment]
@G-Rath in your proposal the same issue you pointed would happen right?
> PRs often end up without complete green tickets so the UI makes it look like its not in a good state to be merged

**Reply by G-Rath** [reply]
No, because the CI would either pass or fail - the app however has a third state: "nothing is semantic".

(See [this comment](https://github.com/octokit/rest.js/discussions/2018#discussioncomment-392076) for more)

**Comment by wolfy1339** [comment]
If we use https://github.com/apps/semantic-pull-requests then changing the PR title would be sufficient to make the test pass

**Reply by oscard0m** [reply]
> If we use [github.com/apps/semantic-pull-requests](https://github.com/apps/semantic-pull-requests) then changing the PR title would be sufficient to make the test pass

Which would be a **desired** or an **undesired** behavior?

**Comment by G-Rath** [comment]
Sorry it's been a while since I've used it so I'm a bit fuzzy on the details, but iirc it was actually about conditions on merging:

![image](https://user-images.githubusercontent.com/3151613/108643271-bfa22a80-750e-11eb-8f00-68a6665b3e74.png)

The app cannot require you to merge or squash a PR, so it can report that a PR is semantic because it's got a single semantic commit which it expects you'll *merge* rather than squash.

Then you squash the PR, making the non-semantic title the commit message and boom you've got a non-semantic message in your branch.

I'm not fussed with what we use - I just found it was a better experience using `commitlint` 🤷

**Reply by wolfy1339** [reply]
You can configure it to require both title and commits.

**Reply by wolfy1339** [reply]
We almost always use the `Squash and merge` function anyways, so the PR title change is sufficient in most cases

**Reply by wolfy1339** [reply]
Alternatively, there is a GitHub actions version: https://github.com/amannn/action-semantic-pull-request

**Reply by oscard0m** [reply]
`validateSingleCommit` would be the option we are looking in case for right?
> When using "Squash and merge" on a PR with only one commit, GitHub
          will suggest using that commit message instead of the PR title for the
          merge commit, and it's easy to commit this by mistake. Enable this option
          to also validate the commit message for one commit PRs.

**Reply by oscard0m** [reply]
What do you think about this Github Action @G-Rath @gr2m ? Do you like it?

**Reply by gr2m** [reply]
> `validateSingleCommit` would be the option we are looking in case for right?
> 
> > When using "Squash and merge" on a PR with only one commit, GitHub
> > will suggest using that commit message instead of the PR title for the
> > merge commit, and it's easy to commit this by mistake. Enable this option
> > to also validate the commit message for one commit PRs.

I've been bothered by GitHub's "squash & merge" behavior if there is only a single commit for a while now. I wanted to build a small GitHub app which adds an empty commit in case a title is changed for a pull request with only a single commit title. 

I think this really should be a built-in feature of https://github.com/apps/semantic-pull-requests, but for the time being I'd just build one myself. 

I'm okay with with installing https://github.com/apps/semantic-pull-requests on the JavaScript @octokit repositories

**Reply by gr2m** [reply]
see also: https://github.com/zeke/semantic-pull-requests/issues/106

**Reply by oscard0m** [reply]
So... until https://github.com/zeke/semantic-pull-requests/issues/106 is not solved, which option do you prefer?
* Would this be an option? https://github.com/amannn/action-semantic-pull-request
* To integrate https://github.com/apps/semantic-pull-requests and wait for the fix with single-commit-pr's?
   * Do you prefer to wait until https://github.com/zeke/semantic-pull-requests/issues/106 is solved?

**Reply by gr2m** [reply]
GitHub Action would be to slow for this use case. I'd build a separate app. We could collaborate on it if you like?

**Reply by oscard0m** [reply]
> GitHub Action would be to slow for this use case. I'd build a separate app. We could collaborate on it if you like?

Count me in!

**Reply by gr2m** [reply]
I invited you to https://github.com/gr2m/squash-commit-app/invitations, lets continue discussion there :)

**Reply by G-Rath** [reply]
> Sorry, we couldn't find that repository invitation. It is possible that the invitation was revoked or that you are not logged into the invited account.
 
😢

**Reply by gr2m** [reply]
@G-Rath Try again :)

**Comment by oscard0m** [comment, marked as answer]
Just to follow up on this. The GitHub app was created 🍾 🍾 🍾 🍾 
https://github.com/apps/squash-commit-app

**Reply by gr2m** [reply]
Great work :)

I've installed it on all relevant @octokit repositories, let's see how it goes! 

I think we would need a slightly different version which takes the semantic commit conventions into account, but let's see how the current version plays out first

**Reply by oscard0m** [reply]
> Great work :)
> 
> I've installed it on all relevant @octokit repositories, let's see how it goes!
> 
> I think we would need a slightly different version which takes the semantic commit conventions into account, but let's see how the current version plays out first

Sure!
